src/model_lifecycle.cc

```diff
@@ -0,0 +1,760 @@
+        std::lock_guard<std::mutex> lock(version_model.second->mtx_);
+        if (version_model.second->state_ == ModelReadyState::READY) {
+          latest = version_model.first;
+          // Tedious, but have to set handle for any "latest" version
+          // at the moment to avoid edge case like the following:
+          // "versions : 1 3 2", version 3 is latest but is requested
+          // to be unloaded when the iterator is examining version 2,
+          // then 'model' will ensure version 3 is still valid
+          *model = version_model.second->model_;
+        }
+      }
+    }
+    if (latest == -1) {
+      return Status(
+          Status::Code::NOT_FOUND,
+          "'" + model_name + "' has no available versions");
+    }
+  } else {
+    std::lock_guard<std::mutex> lock(vit->second->mtx_);
+    if (vit->second->state_ == ModelReadyState::READY) {
+      *model = vit->second->model_;
+    } else {
+      return Status(
+          Status::Code::UNAVAILABLE, "'" + model_name + "' version " +
+                                         std::to_string(version) +
+                                         " is not at ready state");
+    }
+  }
+  return Status::Success;
+}
+
+Status
+ModelLifeCycle::AsyncUnload(const std::string& model_name)
+{
+  LOG_VERBOSE(2) << "AsyncUnload() '" << model_name << "'";
+  std::lock_guard<std::mutex> map_lock(map_mtx_);
+  auto it = map_.find(model_name);
+  if (it == map_.end()) {
+    return Status(
+        Status::Code::INVALID_ARG, "Model to be unloaded has not been served");
+  }
+
+  // Get the existing agent models and notify the unload action
+  const uint64_t now_ns =
+      std::chrono::duration_cast<std::chrono::nanoseconds>(
+          std::chrono::steady_clock::now().time_since_epoch())
+          .count();
+  for (auto& version : it->second) {
+    auto& model_info = version.second;
+    std::lock_guard<std::mutex> lock(model_info->mtx_);
+    model_info->last_update_ns_ = now_ns;
+    // Unload serving model, for model that is in LOADING state,
+    // the updated timestamp will be recognized that there is newer update
+    // on the model info and the load should be aborted
+    if (model_info->state_ == ModelReadyState::READY) {
+      if ((model_info->agent_model_list_ != nullptr) &&
+          (model_info->agent_model_list_->LastActionType() ==
+           TRITONREPOAGENT_ACTION_LOAD_COMPLETE)) {
+        // Only log the error because the model should be unloaded regardless
+        auto status = model_info->agent_model_list_->InvokeAgentModels(
+            TRITONREPOAGENT_ACTION_UNLOAD);
+        if (!status.IsOk()) {
+          LOG_ERROR
+              << "Agent model returns error on TRITONREPOAGENT_ACTION_UNLOAD: "
+              << status.AsString();
+        }
+      }
+
+      // unload
+      model_info->state_ = ModelReadyState::UNLOADING;
+      model_info->state_reason_.clear();
+      model_info->agent_model_list_.reset();
+      model_info->model_.reset();
```

Comment:
Can you add a `ModelInfo::Reset()` or `Unload()` or similar that does all of these actions, and call it throughout the code? This will be easier to maintain if more things need to be done on unload. Seems to happen in a enough places to be worth adding a helper:
- https://github.com/triton-inference-server/core/blob/9676048c25581f58f25a32ecfcbde4ca72f65010/src/model_lifecycle.cc#L420
- https://github.com/triton-inference-server/core/blob/9676048c25581f58f25a32ecfcbde4ca72f65010/src/model_lifecycle.cc#L667
- https://github.com/triton-inference-server/core/blob/9676048c25581f58f25a32ecfcbde4ca72f65010/src/model_lifecycle.cc#L702